TickSpec/FeatureParser.fs

```diff
@@ -1,92 +1,131 @@
-            | scenarioName,tags,steps,Some(exampleTables) ->
-                scenarioName,tags,steps,Some(Array.append exampleTables sharedExamples)
+        // Union tables with the same columns
+        |> Seq.groupBy (fun (h,_) -> h)
+        |> Seq.map (fun (header,tables) ->
+            header, tables |> Seq.collect (fun (_, rows) -> rows) |> Seq.toList)
+        |> Seq.toList
+        // Cross-join tables with different columns
+        |> combinations
+        |> Seq.map processRow
+        |> Seq.choose id
+        |> Seq.groupBy (fun (_,r) -> r)
+        |> Seq.map (fun (row, taggedRows) ->
+            taggedRows
+            |> Seq.fold (fun tags (t,r) ->
+                Seq.append tags t
+            ) Seq.empty
+            |> Seq.distinct
+            |> Seq.toList,
+            row |> Map.toList
         )
 
-/// Parses lines of feature
-let parseFeature (lines:string[]) =
-    let toStep (_,_,_,line,step) = step,line
-    let featureName,background,scenarios,sharedExamples = parseBlocks lines
+    let createStep combination step =
+        let processedStep =
+            match step.Step with
+            | GivenStep s -> replace combination s |> GivenStep
+            | WhenStep s -> replace combination s |> WhenStep
+            | ThenStep s  -> replace combination s |> ThenStep
+
+        let bullets, table, doc =
+            match step.Item with
+            | Some (BulletsItem b) -> Some (b |> List.toArray), None, None
+            | Some (TableItem t) -> None, Some (new Table(t.Header |> List.toArray, t.Rows |> Seq.map List.toArray |> Seq.toArray)), None
+            | Some (DocStringItem d) -> None, None, Some d
+            | None -> None, None, None
+
+        (processedStep, { Number = step.LineNumber; Text=step.LineString; Bullets = bullets; Table = table; Doc = doc})
+
+    let parsedFeatureBlocks = parseBlocks lines
+    let sharedExamples = parsedFeatureBlocks.SharedExamples
+    let background = parsedFeatureBlocks.Background
+
     let scenarios =
-        scenarios
-        |> appendSharedExamples sharedExamples
-        |> Seq.collect (function
-            | name,tags,steps,None ->
+        parsedFeatureBlocks.Scenarios
+        |> Seq.collect (fun scenario ->
+            let examples = scenario.Examples @ sharedExamples
+            let baseTags = parsedFeatureBlocks.Tags @ scenario.Tags
+
+            let exampleCombinations = computeCombinations examples |> Seq.toList
+            let nameFunc =
+                if exampleCombinations.Length > 1 then
+                    fun i -> sprintf "%s (%d)" scenario.Name (i+1)
+                else
+                    fun _ -> scenario.Name
+
+            exampleCombinations
+            |> List.mapi (fun i (tags, combination) ->
+                let name = nameFunc i
                 let steps =
-                    Seq.append background steps
-                    |> Seq.map toStep
+                    background @ scenario.Steps
```

Comment:
I think this will do a List append, Seq.append will have better perf so best to leave

Reply:
You are right.